---
- hosts: "{{ VMware_VM_IP_address }}"
  gather_facts: false
  tasks:
    - block:
        - name: Join VM to the Active Directory domain
          microsoft.ad.membership:
            dns_domain_name: dveurope.dvabb.dev
            domain_admin_user: "{{ AD_USERNAME }}"
            domain_admin_password: "{{ AD_PASSWORD }}"
            hostname: "{{ VMware_VM_name }}"
            domain_ou_path: "OU=ANSI-AUTO,OU=Servers,OU=XE,DC=dveurope,DC=dvabb,DC=dev" 
            state: domain
          register: domain_state

        - name: Show domain state output
          debug: 
            msg: "{{ domain_state }}"

        - name: Reboot the VM if required
          ansible.windows.win_reboot:
          when: domain_state.reboot_required

      rescue:
        - name: Set failure message to be used in the next workflow
          ansible.builtin.set_fact:
            error_message: "Infoblox alloted the IP {{ VMware_VM_IP_address }} for {{ VMware_VM_name }}. Server {{ VMware_VM_name }} built successfully. Server added into ansible inventory. But failure happend while adding the server in domain. Error detail: {{ ansible_failed_result.msg }}"
          when: ansible_failed_result is defined

        - name: Forward failure message
          ansible.builtin.set_stats:
            data:
              error_message: "{{ error_message }}"
          when: ansible_failed_result is defined

        - name: Fail the playbook if there is an error
          ansible.builtin.fail:
            msg: "Playbook failed during execution. Error message: {{ error_message }}"
          when: ansible_failed_result is defined
